sqlx-macros/src/derives/decode.rs

```diff
@@ -0,0 +1,221 @@
+        let generics = &input.generics;
+        let (_, ty_generics, _) = generics.split_for_impl();
+
+        // add db type for impl generics & where clause
+        let mut generics = generics.clone();
+        let predicates = &mut generics.make_where_clause().predicates;
+        for field in fields {
+            let ty = &field.ty;
+            predicates.push(parse_quote!(#ty: sqlx::decode::Decode<sqlx::Postgres>));
+            predicates.push(parse_quote!(sqlx::Postgres: sqlx::types::HasSqlType<#ty>));
+        }
+        let (impl_generics, _, where_clause) = generics.split_for_impl();
+
+        let mut reads: Vec<Vec<Stmt>> = Vec::new();
+        let mut names: Vec<Ident> = Vec::new();
+        for field in fields {
+            let id = &field.ident;
+            names.push(id.clone().unwrap());
+            let ty = &field.ty;
+            reads.push(parse_quote!(
+            if buf.len() < 8 {
+                return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new("Not enough data sent")));
+            }
+
+            let oid = u32::from_be_bytes(std::convert::TryInto::try_into(&buf[0..4]).unwrap());
+            if oid != <sqlx::Postgres as sqlx::types::HasSqlType<#ty>>::type_info().oid() {
+                return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new("Invalid oid")));
+            }
+
+            let len = u32::from_be_bytes(std::convert::TryInto::try_into(&buf[4..8]).unwrap()) as usize;
+
+            if buf.len() < 8 + len {
+                return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new("Not enough data sent")));
+            }
+
+            let raw = &buf[8..8+len];
+            let #id = <#ty as sqlx::decode::Decode<sqlx::Postgres>>::decode(raw)?;
+
+            let buf = &buf[8+len..];
+        ));
+        }
+        let reads = reads.into_iter().flatten();
+
+        tts.extend(quote!(
+        impl #impl_generics sqlx::decode::Decode<sqlx::Postgres> for #ident#ty_generics #where_clause {
+            fn decode(buf: &[u8]) -> std::result::Result<Self, sqlx::decode::DecodeError> {
+                if buf.len() < 4 {
+                    return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new("Not enough data sent")));
+                }
+
+                let column_count = u32::from_be_bytes(std::convert::TryInto::try_into(&buf[..4]).unwrap()) as usize;
+                if column_count != #column_count {
+                    return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new("Invalid column count")));
+                }
+                let buf = &buf[4..];
+
+                #(#reads)*
+
+                if !buf.is_empty() {
+                    return Err(sqlx::decode::DecodeError::Message(std::boxed::Box::new(format!("Too much data sent ({} bytes left)", buf.len()))));
+                }
+
+                Ok(#ident {
+                    #(#names),*
+                })
+            }
+        }
+    ))
+    }
+    Ok(tts)
```

Comment:
Did you and @mehcode already discuss what this should do if the `postgres` feature isn't activated? I would think this should error.

Reply:
no, we did not discuss that yet